### ISSUE TITLE
Support capturing stack traces in Dart

### ISSUE THREAD

**Issue by mprobst** [angular/ts2dart#152]
Dart doesn't provide stack traces by default because apparently capturing stack traces is a severe bottleneck in many applications (ha!). So users have to anticipate that the stack might be used and write:
```catch (ex, stackTrace) { ... }```

For ts2dart, we'd have to come up with a syntax to even express this. Maybe a magic function like so:

```catch (ex) {
  var stackTrace = CAPTURE_DART_STACK(ex.stack);
}```

Which would be a no-op in TypeScript, but translate to the catch statement above for Dart code.

@yjbanov @jbdeboer WDYT?

**Comment by yjbanov** [comment]
Your description is inaccurate. You'll find more here: https://github.com/angular/angular/issues/1407#issuecomment-93918367

tl;dr the `catch(e, s)` syntax allows you to throw any value (including strings, ints) and not lose the stacktrace, which is not possible in JS. That's not your only option, however, as Dart also has identical to JS `Error` class with `stackTrace` field.

**Comment by mprobst** [comment]
Does `Error` always have the stackTrace field set? If so, why don't we just always throw `Error`? That seems to be a best practice in JavaScript, too.

**Comment by yjbanov** [comment]
We might be catching things thrown by the _user_ code (imagine template expression calling into a controller method). Since we don't have control over user code we might be getting errors that are not `Error`. Here's how we could handle it:

Transpile this:

```typescript
try {
   ... call to user code ...
} catch(e) {
  var stack = e.stack;
}
```

To this:

```dart
import 'package:angular2/angular2_dart_runtime.dart';

try {
  ... call to user code ...
} catch(e) {
  var stack = ng2GetStackTrace(e);
}
```

In `angular2_dart_runtime.dart` we'd have:

```dart
/// Extracts stack trace from an error. If [e] is [Error], returns `e.stackTrace`.
/// Otherwise, return current stack trace and issues a warning.
StackTrace ng2GetStackTrace(dynamic e) {
  if (e is Error) {
    return e.stackTrace;
  } else {
    print('WARNING: Angular 2 cannot get stack traces from non-Error objects. '
        'Caught: $e');
    try {
      throw '';
    } catch(_, s) {
      return s;
    }
  }
}
```

This would get you at least an approximation of the stack trace.

**Comment by yjbanov** [comment]
After talking to @jbdeboer, my idea isn't going to be accepted. We have to preserve stacks from non-`Error`s. Your original suggestion sounds reasonable, although I'd use `GET_STACK(ex)`, which means `ex.stack` in JS and `catch(ex, stack)` in Dart. `CAPTURE_DART_STACK` makes it look like we are not interested in capturing the stack in JS.

**Comment by alexeagle** [comment]
I believe this can be closed, now that it's possible to access the stack trace in dart. Just access `e.stack` in TypeScript, and it will be re-written in dart to `e_stack`.

**Comment by jbdeboer** [comment]
At least write a test for the use-case :-)

**Comment by yjbanov** [comment]
Yes, this can be closed. The ts2dart test case is here: https://github.com/angular/ts2dart/pull/153/files#diff-be49c908a2f5e59a5c4373d5459a2130. A test for ng2 will be written in the angular/angular repo as part of https://github.com/angular/angular/issues/1962
